### ISSUE TITLE
Exception when validating field definition when the field type is the actual db does not match the definition

### ISSUE THREAD

**Issue by marinaglancy** [lmscloud-io/moodle-tool_vault#10]
plugin [local_intelliboard](https://moodle.org/plugins/local_intelliboard) defines the field with the type "char" in the install.xml but creates it as "int" in the upgrade.php

To reproduce:
- Install local_intelliboard from the commit 5c1d9698bb14b5598fc84c5e48c2f1afdd87ddbd
- Upgrade local_intelliboard to the current version
- Try to perform a backup in tool_vault

Somehow this plugin itself works but tool_vault was not prepared for the type mismatch and it results in the following exception during database pre-check:

```
Coding error detected, it must be fixed by a programmer: Invalid field definition in table {local_intelliboard_bbb_atten}: XMLDB_TYPE_INTEGER field "userid" has invalid length: 200

#0 /admin/tool/vault/classes/local/xmldb/dbtable.php(74): sql_generator->getFieldSQL()
#1 /admin/tool/vault/classes/local/xmldb/dbtable.php(191): tool_vault\local\xmldb\dbtable->get_field_sql()
#2 /admin/tool/vault/classes/local/xmldb/dbtable.php(169): tool_vault\local\xmldb\dbtable->align_fields_with_defintion()
#3 /admin/tool/vault/classes/local/xmldb/dbstructure.php(212): tool_vault\local\xmldb\dbtable->compare_with_other_table()
#4 /admin/tool/vault/classes/local/xmldb/dbstructure.php(89): tool_vault\local\xmldb\dbstructure->load_actual_tables()
#5 /admin/tool/vault/classes/local/checks/dbstatus.php(48): tool_vault\local\xmldb\dbstructure::load()
#6 /admin/tool/vault/classes/local/checks/check_base.php(211): tool_vault\local\checks\dbstatus->perform()
#7 /admin/tool/vault/classes/local/checks/dbstatus.php(180): tool_vault\local\checks\check_base->execute()
#8 /admin/tool/vault/classes/task/cron_task.php(202): tool_vault\local\checks\dbstatus->execute()
#9 /admin/tool/vault/classes/task/cron_task.php(140): tool_vault\task\cron_task->start_operation()
#10 /lib/classes/cron.php(410): tool_vault\task\cron_task->execute()
#11 /admin/cli/scheduled_task.php(194): core\cron::run_inner_scheduled_task()
#12 {main}
```

**Comment by marinaglancy** [comment]
this is already reported to intelliboard as https://github.com/intelliboard/intelliboard/issues/47

**Comment by marinaglancy** [comment]
released in v2.2 2024-08-19
